File: 0649-Dota2-Senate/src/Solution.java
import java.util.LinkedList;

/**
 * @author qybit
 * @create 2020-12-11 9:28
 * @Since 1.0.0
 */
public class Solution {

    public String predictPartyVictory(String senate) {
        LinkedList<Integer> radiant = new LinkedList<>();
        LinkedList<Integer> dire = new LinkedList<>();
        for (int i = 0; i < senate.length(); i++) {
            char c = senate.charAt(i);
            if (c=='R') {
                radiant.offerLast(i);
            } else {
                dire.offerLast(i);
            }
        }
        int n = senate.length();
        while (!radiant.isEmpty() && !dire.isEmpty()) {
            int r = radiant.pollFirst();
            int d = dire.pollFirst();
            if (r < d) {
                // 叛变了
                radiant.offerLast(r + n);
            } else {
                dire.offerLast(d + n);
            }
        }
        return radiant.isEmpty() ? "Dire" : "Radiant";
    }
}

File: 0649-Dota2-Senate/cpp/solution.cpp
#include "bits/stdc++.h"

using namespace std;

class Solution {
public:
  string predictPartyVictory(string senate) {
    queue<int> r, d;
    int n = senate.size();
    for (int i = 0; i < n; i++) {
      if (senate[i] == 'R')
        r.push(i);
      else
        d.push(i);
    }
    while (r.size() && d.size()) {
      if (r.front() < d.front())
        r.push(r.front() + n);
      else
        d.push(d.front() + n);
      r.pop();
      d.pop();
    }
    if (r.size())
      return "Radiant";
    return "Dire";
  }
};

int main() {}
